NineChronicles.Standalone.Executable/AWSSink.cs

```diff
@@ -0,0 +1,119 @@
+        private readonly CancellationTokenSource _cancellationTokenSource;
+
+        private AsyncProducerConsumerQueue<LogEvent> _queue;
+
+        public AWSSink(AWSCredentials credentials, RegionEndpoint regionEndPoint, Func<string> logGroupNameGetter, Func<string> logStreamNameGetter = null)
+        {
+            Config = new AmazonCloudWatchLogsConfig();
+            Config.RegionEndpoint = regionEndPoint;
+            _queue = new AsyncProducerConsumerQueue<LogEvent>();
+            _client = new AmazonCloudWatchLogsClient(credentials, Config);
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            LogGroupNameGetter = logGroupNameGetter;
+            LogStreamNameGetter = logStreamNameGetter;
+
+            Worker(_cancellationTokenSource.Token);
+        }
+
+        public void Dispose()
+        {
+            _client.Dispose();
+            _cancellationTokenSource.Cancel();
+        }
+
+        public AmazonCloudWatchLogsConfig Config { get; }
+
+        public Func<string> LogGroupNameGetter { get; set; }
+
+        [CanBeNull] public Func<string> LogStreamNameGetter { get; set; }
+
+        public void Emit(LogEvent logEvent)
+        {
+            _queue.Enqueue(logEvent);
+        }
+
+        private async Task Worker(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                while (LogStreamNameGetter is null)
+                {
+                    await Task.Yield();
+                }
+
+                string logGroupName = LogGroupNameGetter();
+                string logStreamName = LogStreamNameGetter();
+                await CreateLogGroup(logGroupName);
+                await CreateLogStreamAsync(logGroupName, logStreamName);
+                string sequenceToken = await GetSequenceToken(logGroupName, logStreamName);
+
+                cancellationToken.ThrowIfCancellationRequested();
+                var logEvent = await _queue.DequeueAsync(cancellationToken);
+                var request = new PutLogEventsRequest(logGroupName, logStreamName, new List<InputLogEvent>
+                {
+                    new InputLogEvent
+                    {
+                        Message = logEvent.RenderMessage(),
+                        Timestamp = logEvent.Timestamp.UtcDateTime,
+                    }
+                });
+                request.SequenceToken = sequenceToken;
+
+                await _client.PutLogEventsAsync(request, cancellationToken);
+            }
+        }
+
+        private async Task<string> GetSequenceToken(string logGroupName, string logStreamName)
+        {
+            var logStreamsResponse = await _client.DescribeLogStreamsAsync(new DescribeLogStreamsRequest(logGroupName)
+            {
+                LogStreamNamePrefix = logStreamName,
+            });
+
+            var stream = logStreamsResponse.LogStreams.First(s => s.LogStreamName == logStreamName);
```

Comment:
Is it okay we assume `logStreamResponse.LogStreams` has elements at least one?

Reply:
I thought it is *yes* because I called `CreateLogGroup()` and `CreateLogStream()` before calling `GetSequenceToken()`. 
```csharp
                await CreateLogGroup(logGroupName);
                await CreateLogStreamAsync(logGroupName, logStreamName);
                string sequenceToken = await GetSequenceToken(logGroupName, logStreamName);
```

Since your review, it seems better to move them here because they are requirements of this function.  I will update it.

Reply:
Applied at 5205285dff715fcb31445150f57c4398d5daa6e3